### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by weiznich** [graphql-rust/juniper#199]
I'm currently in the process of cleaning up the implementation of wundergraph. As part of this I've ported [this graphql orm](https://github.com/hasura/graphql-backend-benchmarks) benchmark to a diesel/wundergraph/juniper/actix_web stack.
This lead to the following results:
![image](https://user-images.githubusercontent.com/1674512/41662707-36d0ecda-7491-11e8-89d1-9ac1af0bd5b9.png)
The corresponding query is
```graphql
query tracks_media_all {
  Tracks {
    id
    name
    media_type_id {
      name
    }
  }
```
which will be translated to the following sql by wundergraph:
```sql
select * from tracks;
select * from media_types where id in (list_of_media_ids_from_tracks);
```

So I started to dig into this. 
After several tries I've used perf to profile the execution of that query. As a result of that I've found out that more then 50% of the execution time is spend in [this function](https://github.com/graphql-rust/juniper/blob/master/juniper/src/types/base.rs#L325). The main issue seems that some `IndexMap` in there gets reconstructed several times and that those reconstructions do not preallocate the corresponding size.

Code: [Wundergraph benchmark](https://github.com/weiznich/wundergraph/tree/infer_schema%2Bbench/wundergraph_bench) (Needs the [postgres variant of the chinook database](https://github.com/xivSolutions/ChinookDb_Pg_Modified/tree/pg_names))

**Comment by LegNeato** [comment]
Thanks for the detailed report! I'll try to repro

**Comment by weiznich** [comment]
I've tried to dig into this a bit further.
The current state are [this commit to juniper](https://github.com/weiznich/juniper/commit/ebe6dd65f49b26f0ca7371776d5ca47a5b6c4824) and [this commit to wundergraph](https://github.com/weiznich/wundergraph/commit/5917fe5b34e2fbc32ee59c585d7ed3bb7c5554c0). Both contain changes that are targeting juniper. That said I'm not sure if everything I've done there is sane or if something will break...
Using this changes it was possible to get on par with hasura at least for 50 and 100 requests per second.

**Comment by LegNeato** [comment]
Obvious thing, but are you testing in release mode with LTO?

**Comment by weiznich** [comment]
> Obvious thing, but are you testing in release mode with LTO?

Yes that's with all optimization enabled. (I enabled even `target-cpu=native`)

That said here are some more information about the test case:
* the test loads ~3500 database entries (tracks) per request. Each track has a corresponding media type (there are 5 types in total). Those types are loaded in a second query. in total there are 2 queries executed.
* the competitor hasura is written in Haskell and translates the graphql request down to a single SQL statement. Even the serialisation is done in postgresql. See this [description](https://blog.hasura.io/architecture-of-a-high-performance-graphql-to-sql-server-58d9944b8a87) for details. In my opinion the advantage of this that most of the performance critical work is done by postgresql and therefore quite optimized. On the other hand this approach does not allow to use a other serialisation format (what graphql allows in theory) or to mix entities existing in postgresql and entitles existing somewhere else (what graphql was designed for)

Given the two patches posted above I'm able to nearly match the performance of hasura for the example query.

**Comment by LegNeato** [comment]
Sorry, didn't have time to set this up today. Is your patch to the test harness public? I wonder if you just switch out the indexmap with another one what the effect would be. It is only used to return fields in the requested order, which while the spec encourages it I am not sure it is required.

**Comment by LegNeato** [comment]
Right, I saw that comment...did I miss the commit to https://github.com/hasura/graphql-backend-benchmarks adding wundergraph?

Thanks again for digging in, hanging in there with us, and writing such a detailed issue!
